### ISSUE TITLE
jobServiceTest.JobServiceTest-testServiceParallelRecursive is flaky.

### ISSUE THREAD

**Issue by DailyDreaming** [DataBiosphere/toil#3127]
Quoting Adam here, we seem to have triggered an intermittent bug via the recursive service test where we get into a situation where a job in the job store wants to be run but the actual job file itself is missing:

https://ucsc-ci.com/databiosphere/toil/-/jobs/56519

```
leader.py                  762 INFO     0 jobs are running, 0 jobs are issued and waiting to run
leader.py                  476 DEBUG    Built the jobs list, currently have 1 jobs to update and 0 jobs issued
leader.py                  406 DEBUG    Updating status of job 'serviceTestParallelRecursive' kind-serviceTestParallelRecursive/instance-dzga18nu with ID kind-serviceTestParallelRecursive/instance-dzga18nu: with result status: 1
singleMachine.py           502 DEBUG    Issuing the command: _toil_worker serviceTestParallelRecursive file:/tmp/toil-test-toil.test.src.jobServiceTest.JobServiceTest-testServiceParallelRecursive-jobstore-3u6lmx0n kind-serviceTestParallelRecursive/instance-dzga18nu with memory: 2147483648, cores: 1, disk: 2147483648
leader.py                  686 INFO     Issued job 'serviceTestParallelRecursive' kind-serviceTestParallelRecursive/instance-dzga18nu with job batch system ID: 1 and cores: 1, disk: 2.0 G, and memory: 2.0 G
singleMachine.py           423 DEBUG    Launched job 1 as child 6933
singleMachine.py           472 ERROR    Got exit code 1 (indicating failure) from job _toil_worker serviceTestParallelRecursive file:/tmp/toil-test-toil.test.src.jobServiceTest.JobServiceTest-testServiceParallelRecursive-jobstore-3u6lmx0n kind-serviceTestParallelRecursive/instance-dzga18nu.
singleMachine.py           483 DEBUG    Child 6933 for job 1 succeeded
singleMachine.py           578 DEBUG    Ran jobID: 1 with exit value: 1
leader.py                  524 WARNING  Job failed with exit value 1: 'serviceTestParallelRecursive' kind-serviceTestParallelRecursive/instance-dzga18nu
leader.py                  895 DEBUG    Job 'serviceTestParallelRecursive' kind-serviceTestParallelRecursive/instance-dzga18nu continues to exist (i.e. has more to do)
statsAndLogging.py          91 WARNING  The job seems to have left a log file, indicating failure: 'serviceTestParallelRecursive' kind-serviceTestParallelRecursive/instance-dzga18nu
statsAndLogging.py          94 WARNING  Log from job kind-serviceTestParallelRecursive/instance-dzga18nu follows:
=========>
	[2020-07-29T17:12:20+0000] [MainThread] [I] [toil.worker] ---TOIL WORKER OUTPUT LOG---
	[2020-07-29T17:12:20+0000] [MainThread] [I] [toil] Running Toil version 4.2.0a1-1551fe89bcec386859885620ec772ef705e6c7c7 on host runner-2j87tsx-project-3-concurrent-9lnkp6.
	[2020-07-29T17:12:20+0000] [MainThread] [D] [toil] Configuration: {'workflowID': 'ba4e1c02-3000-478b-8fff-60c168add778', 'workflowAttemptNumber': 7, 'jobStore': 'file:/tmp/toil-test-toil.test.src.jobServiceTest.JobServiceTest-testServiceParallelRecursive-jobstore-3u6lmx0n', 'logLevel': 'DEBUG', 'workDir': None, 'noStdOutErr': False, 'stats': False, 'clean': 'onSuccess', 'cleanWorkDir': 'always', 'clusterStats': None, 'restart': True, 'batchSystem': 'singleMachine', 'disableAutoDeployment': False, 'environment': {}, 'statePollingWait': 1, 'maxLocalJobs': 4, 'manualMemArgs': False, 'parasolCommand': 'parasol', 'parasolMaxBatches': 10000, 'scale': 1, 'linkImports': False, 'moveExports': False, 'mesosMasterAddress': '10.244.2.209:5050', 'kubernetesHostPath': None, 'provisioner': None, 'nodeTypes': [], 'nodeOptions': None, 'minNodes': None, 'maxNodes': [10], 'targetTime': 1800, 'betaInertia': 0.1, 'scaleInterval': 60, 'preemptableCompensation': 0.0, 'nodeStorage': 50, 'nodeStorageOverrides': [], 'metrics': False, 'maxPreemptableServiceJobs': 9223372036854775807, 'maxServiceJobs': 9223372036854775807, 'deadlockWait': 60, 'deadlockCheckInterval': 30, 'defaultMemory': 2147483648, 'defaultCores': 1, 'defaultDisk': 2147483648, 'readGlobalFileMutableByDefault': False, 'defaultPreemptable': False, 'maxCores': 9223372036854775807, 'maxMemory': 9223372036854775807, 'maxDisk': 9223372036854775807, 'retryCount': 2, 'enableUnlimitedPreemptableRetries': False, 'maxJobDuration': 9223372036854775807, 'rescueJobsFrequency': 3600, 'disableCaching': True, 'disableChaining': False, 'disableJobStoreChecksumVerification': False, 'maxLogFileSize': 64000, 'writeLogs': None, 'writeLogsGzip': None, 'writeLogsFromAllJobs': False, 'sseKey': None, 'servicePollingInterval': 1.0, 'useAsync': True, 'forceDockerAppliance': False, 'runCwlInternalJobsOnWorkers': False, 'statusWait': 3600, 'disableProgress': False, 'debugWorker': False, 'disableWorkerOutputCapture': False, 'badWorker': 0.5, 'badWorkerFailInterval': 0.1, 'cwl': False}
	[2020-07-29T17:12:20+0000] [MainThread] [D] [toil.deferred] Running for file /tmp/node-ba4e1c02-3000-478b-8fff-60c168add778-50e3d6f921efafe4b3a16de05e0e88db/deferred/func97lm_poq
	[2020-07-29T17:12:20+0000] [MainThread] [D] [toil.worker] Parsed job wrapper
	[2020-07-29T17:12:20+0000] [MainThread] [D] [toil.worker] Job is a checkpoint
	[2020-07-29T17:12:20+0000] [MainThread] [D] [toil.worker] The checkpoint jobs seems to have completed okay, removing any checkpoint files to delete.
	[2020-07-29T17:12:20+0000] [MainThread] [D] [toil.worker] Got a command to run: _toil files/for-job/kind-serviceTestParallelRecursive/instance-dzga18nu/cleanup/file-sy2anui3/stream /builds/databiosphere/toil/src toil.test.src.jobServiceTest False
	[2020-07-29T17:12:20+0000] [MainThread] [D] [toil.job] Loading user module ModuleDescriptor(dirPath='/builds/databiosphere/toil/src', name='toil.test.src.jobServiceTest', fromVirtualEnv=False).
	Traceback (most recent call last):
	  File "/builds/databiosphere/toil/src/toil/worker.py", line 352, in workerScript
	    job = Job._loadJob(jobGraph.command, jobStore)
	  File "/builds/databiosphere/toil/src/toil/job.py", line 940, in _loadJob
	    jobStore.readFile(pickleFile, filename)
	  File "/builds/databiosphere/toil/src/toil/jobStores/fileJobStore.py", line 415, in readFile
	    self._checkJobStoreFileID(jobStoreFileID)
	  File "/builds/databiosphere/toil/src/toil/jobStores/fileJobStore.py", line 686, in _checkJobStoreFileID
	    raise NoSuchFileException(jobStoreFileID)
	toil.jobStores.abstractJobStore.NoSuchFileException: File 'files/for-job/kind-serviceTestParallelRecursive/instance-dzga18nu/cleanup/file-sy2anui3/stream' does not exist.
	[2020-07-29T17:12:20+0000] [MainThread] [E] [toil.worker] Exiting the worker because of a failed job on host runner-2j87tsx-project-3-concurrent-9lnkp6
<=========
jobGraph.py                147 WARNING  Due to failure we are reducing the remaining retry count of job 'serviceTestParallelRecursive' kind-serviceTestParallelRecursive/instance-dzga18nu with ID kind-serviceTestParallelRecursive/instance-dzga18nu to 1
leader.py                  975 DEBUG    Added job: 'serviceTestParallelRecursive' kind-serviceTestParallelRecursive/instance-dzga18nu to active jobs
leader.py                  476 DEBUG    Built the jobs list, currently have 1 jobs to update and 0 jobs issued
```

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-592)
┆Issue Number: TOIL-592

**Comment by adamnovak** [comment]
Here's another example: https://ucsc-ci.com/databiosphere/toil/-/jobs/56875

This seems to happen quite often; we must have introduced it recently.

**Comment by adamnovak** [comment]
I bet this is caused by the update deduplication changes in https://github.com/DataBiosphere/toil/pull/3108 somehow. We're probably losing an update that says that the whole job that has services is done, and we're trying to schedule it again when we shouldn't, somehow.

@DailyDreaming, did you ever write up that architecture doc that explains why we have Job, BaseJob, JobNode, JobGraph, and how they're all supposed to work together? Right now I'm inclined to start rewriting core parts of that system because I don't understand them, but that can't be a good idea.

It looks to me like we started with Job, then decided we needed JobGraph (which is really a graph node) so we could think about scheduling without deserializing the whole Job, and so duplicated all the scheduling dependency information into JobGraph. Then somehow we decided we needed a common way to talk about Job and JobGraph objects, and instead of using BaseJob we made a JobNode class between BaseJob and JobGraph, and made it constructable from either a Job or a JobGraph.

Then we invented services and checkpoints and tacked on a bunch of service-related properties and made a ServiceJobNode.

We are in trouble now because we have about 3 copies of every important piece of job-related information (the Job itself, the JobGraph we pass around to talk about it, and random fields in the leader/ToilState holding return values), and we don't really know what all of them mean (because JobGraph and JobNode don't quite document their fields).

I think we have to slim this down somehow. JobGraph and JobNode and BaseJob should all be combined into one class (JobDescription?), the purpose of which is to represent everything you need to know about a job and how it relates to other jobs, without the user data (arguments, return values, user-added fields, etc.) needed to actually run it. Job should extend that with the actual job-running info. We should stop just pure-pickling everything and leave it up to the JobStore how to store things so that, given a job ID, you can retrieve either the JobDescription or the actual invoke-able Job, and have a guarantee that they agree with each other about relationships to other jobs. All references to other jobs should be by ID; no more holding full Jobs in lists inside other jobs. (We might need to do something special here to support the way you can just instantiate jobs and wire them up, maybe a little cache that will get written back to the JobStore, but that should be special and not the ordinary way that jobs refer to each other.)

Modifications to existing jobs ought to go through JobStore methods, or maybe some kind of commit() method on the Job/JobDescription. We shouldn't be fiddling with the writable fields of these things so freely all over the codebase.

**Comment by adamnovak** [comment]
Absence of the job is supposed to indicate completion no matter what the batch system said.

**Comment by adamnovak** [comment]
I haven't seen this happening since the job refactor; @DailyDreaming have you been restarting tests that fail like this?

**Comment by adamnovak** [comment]
I have determined that this issue is fixed; none of the `py37_main` tests that failed had this complaint in them since we merged #3250.

```
rm -f bad_ids.txt
for PAGE in {1..20} ; do
    curl -sSL --header "PRIVATE-TOKEN: ${GITLAB_ACCESS_TOKEN}" 'https://ucsc-ci.com/api/v4/projects/3/jobs?scope[]=failed&per_page=100&page='${PAGE} | jq '.[] | select(.name == "py37_main") | .id | select(. >= 56875)' >>bad_ids.txt
done

wc -l bad_ids.txt

for JOB_ID in $(cat bad_ids.txt) ; do
    curl -sSL --header "PRIVATE-TOKEN: ${GITLAB_ACCESS_TOKEN}" 'https://ucsc-ci.com/api/v4/projects/3/jobs/'${JOB_ID}'/trace' >log.txt
    grep "NoSuchFileException: File 'files/for-job/kind-serviceTestParallelRecursive" log.txt
    if [[ $? == 0 ]] ; then
        echo "*** Bad Job: ${JOB_ID}"
    else
        echo "OK: ${JOB_ID}"
    fi
done
```

The most recent occurrence of this was in job 59960 in September 2020, and the merge was in October 2020.
